### ISSUE TITLE
Blocks are overlapping after full compaction

### ISSUE THREAD

**Issue by hpbieker** [influxdata/influxdb#9255]
### Bug report

__System info:__ 
InfluxDB 1.3.7, Ubuntu 16.4, 128 GB RAM, 16 cores.

__Steps to reproduce:__

Make sure a "full compaction" of the shard has ran.
```
November 3rd 2017, 17:04:13.000	compacted full 4 files into 3 files in 15m46.744294449s engine=tsm1
November 3rd 2017, 17:04:13.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/5269/000001594-000000004.tsm.tmp (#0) engine=tsm1
November 3rd 2017, 17:04:13.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/5269/000001594-000000005.tsm.tmp (#1) engine=tsm1
November 3rd 2017, 17:04:13.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/5269/000001594-000000006.tsm.tmp (#2) engine=tsm1
November 3rd 2017, 16:48:27.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/5269/000001594-000000002.tsm.tmp (#0) engine=tsm1
November 3rd 2017, 16:48:27.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000000886-000000005.tsm (#0) engine=tsm1
November 3rd 2017, 16:48:27.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001594-000000002.tsm (#2) engine=tsm1
November 3rd 2017, 16:48:27.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/5269/000001594-000000003.tsm.tmp (#1) engine=tsm1
November 3rd 2017, 16:48:27.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000000886-000000006.tsm (#1) engine=tsm1
November 3rd 2017, 16:48:27.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001594-000000003.tsm (#3) engine=tsm1
November 3rd 2017, 16:41:35.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001591-000000003.tsm (#4) engine=tsm1
November 3rd 2017, 16:41:35.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001594-000000001.tsm (#6) engine=tsm1
November 3rd 2017, 16:41:35.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001337-000000005.tsm (#0) engine=tsm1
November 3rd 2017, 16:41:35.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001579-000000005.tsm (#1) engine=tsm1
November 3rd 2017, 16:41:35.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001587-000000003.tsm (#3) engine=tsm1
November 3rd 2017, 16:41:35.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001583-000000003.tsm (#2) engine=tsm1
November 3rd 2017, 16:41:35.000	compacting full group (0) /media/db_volume/data/XXX/autogen/5269/000001593-000000002.tsm (#5) engine=tsm1
```

When I then run influx_inspect dumptsm, I get the following suspicious data:
```
  390166        2017-10-23T23:49:18.069Z        2017-10-24T00:03:52.359Z        459715575       3360    XXX       value
  390167        2017-10-24T00:03:53.359Z        2017-10-29T23:55:45.973Z        459718935       4410    XXX       value
  390168        2017-10-24T00:14:47.679Z        2017-10-24T00:29:26.606Z        459723345       3664    XXX       value
```

As you can see, the block in the midle (390167) almost spans six days, covering multiple other blocks. However, the third block (390168) is just a few minutes after the first block. To me it looks like the full compaction is not done properly. It looks like this is causing the OOM in https://github.com/influxdata/influxdb/issues/9241 . It is unclear if the problem is that the block has the wrong index or if the block actually contains values that should have been moved to other blocks.

The output of command below is available here: https://gist.github.com/hpbieker/74c758324f9bcc257fa1b52a76df6bd2
```
$ influx_inspect dumptsm -index -filter-key XXX testdata/data/XXX/autogen/5269/000001594-000000004.tsm
```

__Expected behavior:__ 
Full compaction should always give non-overlapping blocks for a given measurement/field/tag combination.

__Actual behavior:__ 
I get overlapping blocks.

__Additional info:__ 
I just had a look at the query logs, and identified some slow queries for short time periods. Then I ran the command as above for the shard, and found the same problem (end time of third block) -- overlapping blocks in the shards:

```
$ influx_inspect dumptsm -index -filter-key XXX /media/db_volume/data/XXX/autogen/4863/000001586-000000006.tsm
  File: /media/db_volume/data/XXX/autogen/4863/000001586-000000006.tsm
  Time Range: 2017-10-09T00:00:00Z - 2017-10-15T23:59:59.994Z
  Duration: 167h59m59.994s   Series: 58423   File Size: 2147544332
  Pos           Min Time                Max Time                Ofs             Size    Key                             Field
[..]
  489838        2017-10-10T23:29:59.347Z        2017-10-10T23:46:38.347Z        1336452239      30      XXX       someField
  489839        2017-10-10T23:46:39.347Z        2017-10-15T23:51:01.552Z        1336452269      1372    XXX       someField
  489840        2017-10-11T00:00:00.347Z        2017-10-11T00:16:39.347Z        1336380574      30      XXX       someField
[..]
```
This is the details from the problem block 489839 (which I belive is listed as 489838 below due to an offset error in influx_inspect https://github.com/influxdata/influxdb/issues/9254).
```
Blocks:
  Blk           Chk             Ofs             Len     Type    Min Time                        Points  Enc [T/V]       Len [T/V]
  489837        3008412047      356568942287491 26      int64   2017-10-10T23:29:59.347Z        1000    rle/rle         12/12
  489838        2469651021      356570278674727 1368    int64   2017-10-10T23:46:39.347Z        1000    s8b/rle         1353/12
  489839        1126674234      356571615063335 26      int64   2017-10-11T00:00:00.347Z        1000    rle/rle         12/12
  489840        2640338492      356572951451973 26      int64   2017-10-11T00:16:40.347Z        1000    rle/rle         12/12
```

Too me it looks like there is something wrong with the full compaction/writer code that causes overlapping blocks with overlapping time periods for same measurement/field to be written.

As you can see from the log below, the files where fully compacted and it was written with Influx 1.3.7:
```
November 7th 2017, 11:18:34.000	InfluxDB starting, version 1.3.7, branch 1.3, commit 2d474a3089bcfce6b472779be9470a1f0ef3d5e4
November 3rd 2017, 19:06:16.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/4863/000001586-000000007.tsm.tmp (#3) engine=tsm1
November 3rd 2017, 19:06:16.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/4863/000001586-000000005.tsm.tmp (#1) engine=tsm1
November 3rd 2017, 19:06:16.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/4863/000001586-000000004.tsm.tmp (#0) engine=tsm1
November 3rd 2017, 19:06:16.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/4863/000001586-000000006.tsm.tmp (#2) engine=tsm1
November 3rd 2017, 18:50:52.000	compacting full group (0) /media/db_volume/data/XXX/autogen/4863/000001586-000000003.tsm (#3) engine=tsm1
November 3rd 2017, 18:50:52.000	compacted full group (0) into /media/db_volume/data/XXX/autogen/4863/000001586-000000003.tsm.tmp (#0) engine=tsm1
November 3rd 2017, 18:47:34.000	compacting full group (0) /media/db_volume/data/XXX/autogen/4863/000001586-000000002.tsm (#3) engine=tsm1
November 3rd 2017, 14:47:35.000	compacted level 1 group (0) into /media/db_volume/data/XXX/autogen/4863/000001586-000000002.tsm.tmp (#0) engine=tsm1
November 3rd 2017, 14:47:35.000	compacting level 1 group (0) /media/db_volume/data/XXX/autogen/4863/000001586-000000001.tsm (#1) engine=tsm1
November 2nd 2017, 15:22:00.000	InfluxDB starting, version 1.3.7, branch 1.3, commit 2d474a3089bcfce6b472779be9470a1f0ef3d5e4
```

**Comment by hpbieker** [comment]
I just wrote the test below, and it fails on master / 1.4 and 1.3 branches.

It creates two tsm files:
* 1st with timestamps [4, 5, 7]
* 2nd with timestamps [1, 2, 3]

After compacting, the timestamps are [1, 2, 3, 7, 4, 5] and not [1, 2, 3, 4, 5, 7] which is exepected.

It seems to be a bug in the merging where it is assumed that newer tsm files has more recent values than older tsm files? Can you please verify, @jwilder or @benbjohnson ?

```
$ go test -v -run ^TestCompactor_Compact_WrongOrderBlocks$
=== RUN   TestCompactor_Compact_WrongOrderBlocks
--- FAIL: TestCompactor_Compact_WrongOrderBlocks (0.02s)
        compact_test.go:2515: time mismatch: got 7, exp 4
FAIL
exit status 1
FAIL    github.com/influxdata/influxdb/tsdb/engine/tsm1 0.120s
```

The test looks like this:

```
$ git diff hpb-compact-wrong-order-blocks 1.3|cat
diff --git a/tsdb/engine/tsm1/compact_test.go b/tsdb/engine/tsm1/compact_test.go
index 31fdee19d..a339f523a 100644
--- a/tsdb/engine/tsm1/compact_test.go
+++ b/tsdb/engine/tsm1/compact_test.go
@@ -262,6 +262,75 @@ func TestCompactor_Compact_OverlappingBlocks(t *testing.T) {
 }

 // Ensures that a compaction will properly merge multiple TSM files
+func TestCompactor_Compact_WrongOrderBlocks(t *testing.T) {
+       dir := MustTempDir()
+       defer os.RemoveAll(dir)
+
+       // write 2 TSM files with different data and one new point
+       a1 := tsm1.NewValue(4, 1.1)
+       a2 := tsm1.NewValue(5, 1.1)
+       a3 := tsm1.NewValue(7, 1.1)
+
+       writes := map[string][]tsm1.Value{
+               "cpu,host=A#!~#value": []tsm1.Value{a1, a2, a3},
+       }
+       f1 := MustWriteTSM(dir, 1, writes)
+
+       b1 := tsm1.NewValue(1, 1.2)
+       b2 := tsm1.NewValue(2, 1.2)
+       b3 := tsm1.NewValue(3, 1.2)
+
+       writes = map[string][]tsm1.Value{
+               "cpu,host=A#!~#value": []tsm1.Value{b1, b2, b3},
+       }
+       f2 := MustWriteTSM(dir, 3, writes)
+
+       compactor := &tsm1.Compactor{
+               Dir:       dir,
+               FileStore: &fakeFileStore{},
+               Size:      2,
+       }
+
+       compactor.Open()
+
+       files, err := compactor.CompactFast([]string{f1, f2})
+       if err != nil {
+               t.Fatalf("unexpected error writing snapshot: %v", err)
+       }
+
+       if got, exp := len(files), 1; got != exp {
+               t.Fatalf("files length mismatch: got %v, exp %v", got, exp)
+       }
+
+       r := MustOpenTSMReader(files[0])
+
+       if got, exp := r.KeyCount(), 1; got != exp {
+               t.Fatalf("keys length mismatch: got %v, exp %v", got, exp)
+       }
+
+       var data = []struct {
+               key    string
+               points []tsm1.Value
+       }{
+               {"cpu,host=A#!~#value", []tsm1.Value{b1, b2, b3, a1, a2, a3}},
+       }
+
+       for _, p := range data {
+               values, err := r.ReadAll(p.key)
+               if err != nil {
+                       t.Fatalf("unexpected error reading: %v", err)
+               }
+
+               if got, exp := len(values), len(p.points); got != exp {
+                       t.Fatalf("values length mismatch %s: got %v, exp %v", p.key, got, exp)
+               }
+
+               for i, point := range p.points {
+                       assertValueEqual(t, values[i], point)
+               }
+       }
+}
+
 func TestCompactor_Compact_OverlappingBlocksMultiple(t *testing.T) {
        dir := MustTempDir()
        defer os.RemoveAll(dir)
```

**Comment by jwilder** [comment]
Fixed via #9267
